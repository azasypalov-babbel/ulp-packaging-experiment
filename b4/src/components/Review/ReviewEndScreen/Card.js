import React from 'react';
import PropTypes from 'prop-types';
import cx from '../../../lib/cxHelper';

import ReviewItem from './ReviewItem';

const delay = 400;

class Card extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      currentIndex: 0
    };
    this.timers = [];
  }

  componentDidMount() {
    this.timers = this.props.itemsList.map((item, index, array) =>
      setTimeout(() => {
        this.setState({ currentIndex: index + 1 }, this.updateItemListScroll);
        if (index === array.length - 1) {
          this.props.onAnimationComplete();
        }
      }, delay + (delay * index), index));
  }

  updateItemListScroll() {
    this.reviewItemListRef.scrollTop = this.reviewItemListRef.scrollHeight;
  }

  componentWillUnmount() {
    this.timers.map(clearTimeout);
  }

  render() {
    const { type, itemsList, onItemClick, title } = this.props;
    const cardType = type  === 'correct' ? 'card--correct' : 'card--incorrect';
    const dataReviewType = type  === 'correct' ? 'item-correct' : 'item-incorrect';

    return (
      <div className={cx('card', cardType)}>
        <h2 className={cx('card__title')}>{title}</h2>
        <div className={cx('card__body')}>
          <ul ref={(el) => { this.reviewItemListRef = el; }} className={cx('review-item-list')}>
            {itemsList.slice(0, this.state.currentIndex).map((item, i) => (
              <li key={i}>
                <ReviewItem data-review={dataReviewType} item={item} onItemClick={onItemClick} />
              </li>
            ))}
          </ul>
        </div>
      </div>
    );
  }
}

Card.propTypes = {
  type: PropTypes.string.isRequired,
  itemsList: PropTypes.arrayOf(PropTypes.shape({
    learnLanguageText: PropTypes.string.isRequired,
    sound: PropTypes.shape({
      id: PropTypes.string.isRequired
    }).isRequired
  })).isRequired,
  onItemClick: PropTypes.func.isRequired,
  onAnimationComplete: PropTypes.func,
  title: PropTypes.string.isRequired
};

Card.defaultProps = {
  onAnimationComplete: () => {}
};

export default Card;
