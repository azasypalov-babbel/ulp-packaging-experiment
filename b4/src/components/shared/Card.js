import { transparentize, rem } from 'polished';
import styled, { css } from 'styled-components';

const Card = styled.div`
  box-sizing: border-box;
  overflow: hidden;

  background: ${(props) => props.theme.cascada.white};

  border-radius: ${rem('4px')};

  transition: box-shadow 0.3s ease;

  box-shadow: 0 ${rem('2px')} ${rem('20px')} 0 ${(props) => transparentize(0.9, props.theme.cascada.spaceGrayW15)};

  outline: none;

  ${(props) => Boolean(props.onClick) && css`

    cursor: pointer;

    &:hover, &:focus {
      box-shadow: 0 ${rem('8px')} ${rem('40px')} 0 ${(props) => transparentize(0.8, props.theme.cascada.spaceGrayW15)};
    }

    &:active {
      box-shadow: 0 ${rem('2px')} ${rem('10px')} 0 ${(props) => transparentize(0.7, props.theme.cascada.spaceGrayW15)};
    }
  `}
`;

export default Card;
