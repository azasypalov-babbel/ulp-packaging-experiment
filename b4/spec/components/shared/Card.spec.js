import React from 'react';
import Card from '../../../src/components/shared/Card';
import { mountWithTheme } from './themeMock';

describe('<Card />', () => {
  test('it renders', () => {
    const wrapper = mountWithTheme(<Card />);
    expect(wrapper).toMatchSnapshot();
  });
});
