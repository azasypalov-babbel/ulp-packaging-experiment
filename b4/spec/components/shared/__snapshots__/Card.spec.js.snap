// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> it renders 1`] = `
.c0 {
  box-sizing: border-box;
  overflow: hidden;
  background: #ffffff;
  border-radius: rem(4px);
  -webkit-transition: box-shadow 0.3s ease;
  transition: box-shadow 0.3s ease;
  box-shadow: 0 rem(2px) rem(20px) 0 transparentize(0.9,#565d64);
  outline: none;
}

<styled.div>
  <div
    className="c0"
  />
</styled.div>
`;
